tagstudio/src/core/library/alchemy/library.py

```diff
@@ -0,0 +1,769 @@
+    selectinload,
+    make_transient,
+)
+
+from .db import make_tables
+from .enums import TagColor, FilterState
+from .fields import (
+    DatetimeField,
+    Field,
+    TagBoxField,
+    TagBoxTypes,
+    TextField,
+    DefaultField,
+    FieldID,
+)
+from .joins import TagSubtag, TagField
+from .models import Entry, Preferences, Tag, TagAlias
+from ...constants import LibraryPrefs, TS_FOLDER_NAME, TAG_ARCHIVED, TAG_FAVORITE
+
+LIBRARY_FILENAME: str = "ts_library.sqlite"
+
+logger = structlog.get_logger(__name__)
+
+import re
+import unicodedata
+
+
+def slugify(input_string: str) -> str:
+    # Convert to lowercase and normalize unicode characters
+    slug = unicodedata.normalize("NFKD", input_string.lower())
+
+    # Remove non-word characters (except hyphens and spaces)
+    slug = re.sub(r"[^\w\s-]", "", slug).strip()
+
+    # Replace spaces with hyphens
+    slug = re.sub(r"[-\s]+", "-", slug)
+
+    return slug
+
+
+def get_default_tags() -> list[Tag]:
+    archive_tag = Tag(
+        id=TAG_ARCHIVED,
+        name="Archived",
+        aliases={TagAlias(name="Archive")},
+        color=TagColor.RED,
+    )
+
+    favorite_tag = Tag(
+        id=TAG_FAVORITE,
+        name="Favorite",
+        aliases={
+            TagAlias(name="Favorited"),
+            TagAlias(name="Favorites"),
+        },
+        color=TagColor.YELLOW,
+    )
+
+    return [archive_tag, favorite_tag]
+
+
+class Library:
+    """Class for the Library object, and all CRUD operations made upon it."""
+
+    library_dir: Path | str
+    missing_files: list[str]
+    dupe_files: list[str]
+    engine: Engine | None
+    dupe_entries: list[Entry]  # TODO
+
+    def __init__(self):
+        self.clear_internal_vars()
+
+    def open_library(self, library_dir: Path | str) -> None:
+        self.clear_internal_vars()
+
+        if library_dir == ":memory:":
+            connection_string = f"sqlite:///{library_dir}"
+            self.library_dir = library_dir
+        else:
+            self.library_dir = Path(library_dir)
+            self.verify_ts_folders(self.library_dir)
+
+            connection_string = (
+                f"sqlite:///{self.library_dir / TS_FOLDER_NAME / LIBRARY_FILENAME}"
+            )
+
+        logger.info("opening library", connection_string=connection_string)
+        self.engine = create_engine(connection_string)
+        with Session(self.engine) as session:
+            make_tables(self.engine)
+
+            tags = get_default_tags()
+            try:
+                session.add_all(tags)
+                session.commit()
+            except IntegrityError:
+                # default tags may exist already
+                session.rollback()
+
+            for pref in LibraryPrefs:
+                try:
+                    session.add(Preferences(key=pref.name, value=pref.value))
+                    session.commit()
+                except IntegrityError:
+                    logger.error("pref already exists", pref=pref)
```

Comment:
This errors out twice everytime a library is opened. Would is make more sense to not have an error message like adding the default tags above? or adding some sort of check to see if they default stuff is missing?